temporal-sdk/src/main/java/io/temporal/nexus/WorkflowRunNexusOperationHandler.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.nexus;
+
+import io.nexusrpc.handler.*;
+import io.nexusrpc.handler.OperationHandler;
+import io.temporal.client.WorkflowClient;
+import io.temporal.common.Experimental;
+
+/**
+ * WorkflowRunNexusOperationHandler can be used to create OperationHandlers that will trigger a
+ * workflow run
+ */
+@Experimental
+public final class WorkflowRunNexusOperationHandler {
```

Comment:
I wonder if this should just move to `OperationHandler`. Ideally there are not too many static helpers distributed amongst several static-method-only class holders. Maybe it'd be clearer to just expose `RunWorkflowOperation` as `WorkflowRunOperationHandler` and move some of this to that? Can still use static instead w/ private constructor if you're really not wanting to expose the constructor.

Reply:
So we [rename](https://github.com/temporalio/sdk-java/pull/2231#discussion_r1771305064) `OperationHandler`  to `WorkflowClientOperationHandler` then move  `fromWorkflowMethod` and `fromWorkflowHandle` onto `WorkflowClientOperationHandler`?

Reply:
Thinking about this a bit, I think you could have:

* `WorkflowClientOperationHandler` - Handler for operations needing a Temporal client
  * Ideally it's an abstract class that implements `OperationHandler` and provides a `getTemporalClient`
    * But I understand if we want to hide async operations being built by users with Temporal clients, so can make the abstract class empty with a package-private constructor
  * Has the `sync` call that accepts a client
* `WorkflowRunOperationHandler` - Handler for operations based on starting a workflow run
  * Extends `WorkflowClientOperationHandler`
  * Is basically this class, so has `fromWorkflowMethod` and `fromWorkflowHandle`
  * Implementation is that of `RunWorkflowOperation`, you don't need that separate (can still make constructor private for now though since you have these static helpers)
  * If there are concerns of making this class too visible, could make this package private and expose `fromWorkflowMethod` and `fromWorkflowHandle` on the exposed `WorkflowClientOperationHandler`

I think this matches what our original Nexus thoughts were that Go couldn't express: A workflow-run handler is an instance of a Temporal client handler which is an instance of a generic Nexus handler. And you cut 3 classes to 2 and get rid of static-method-only classes here.

I would expect later there will be a `WorkflowUpdateOperationHandler` or at least a `fromWorkflowUpdate`-type of thing that will leverage `WorkflowClientOperationHandler`.

Reply:
I would strongly prefer to stick to the proposed design, if we want to rename `WorkflowRunNexusOperationHandler` to `WorkflowRunOperationHandler` I am fine with that

Reply:
Can you clarify a bit? We shouldn't call something a handler when it's not IMO. To have a simple Temporal client handler for Temporal client needs and a more specific version of that for workflow run seems very reasonable.

Reply:
`WorkflowRunNexusOperationHandler ` helps create nexus operation handlers, that is why I added `OperationHandler` , I like this name, but if you feel strongly about a different name though suggestions welcome. As [discussed](https://www.notion.so/temporalio/Temporal-Java-SDK-workflow-to-operation-2c8d5a4975574116b4df9fcd7f9d7e73?pvs=4#3aa94e554f524319babafecf47ab64be) in the proposal we do not need to add Temporal specific operation handlers at this time.

Reply:
Is there any harm in making it `final` and moving the implementation from `RunWorkflowOperation` to here? At least then it actually is an operation handler and you've reduced the class size down. The constructor can remain package private or private, so it is effectively the same from a user POV except it's clear what it is conceptually.

Reply:
Yes I do not want to expose the implementation details here and keep the concepts and class count for users simpler. The important part of this API is it returns an `OperationHandler ` they can register with their worker.

>POV except it's clear what it is conceptually

From my perspective returning some Java SDK specific handler is more confusing then returning a standard nexus handler.

Reply:
If we want to expose `RunWorkflowOperation` in the future we can  (though I am not convinced it would ever provide value to users), but at this time, as we discussed in the proposal, I am against exposing more concepts with no value to the user.

Reply:
Would you at least be willing to combine the 1 static method from `WorkflowClientOperationHandler` and the 2 from `WorkflowRunNexusOperationHandler` to a single static-method-only class called `OperationHandlers` or something? I know we haven't done plural in Java SDK with other things, but it is fairly common in Java standard library.

Reply:
Yes that is what I was asking [here](https://github.com/temporalio/sdk-java/pull/2231#discussion_r1771729629) , I wouldn't want to call it `OperationHandlers` though as I don't like the plural and I don't like the similarity with `OperationHandler` from the Nexus SDK. If we want to call it something like `TemporalOperationHandler` I'd be OK with that.

Reply:
Is it possible to name it _anything_ but an operation handler if it is not an operation handler?

Reply:
:+1: I do not have a super strong opinion, but I do prefer not naming it operation handler if it's not (maybe in the future there is a `WorkflowClientOperationHandler`) unless maybe in the future it does become one.

Will mark approved as is, but yes, I do prefer `TemporalOperationHandlers` or even `WorkflowClientOperationHandlers`.

Reply:
Moved these methods to `WorkflowClientOperationHandlers `

https://github.com/temporalio/sdk-java/pull/2231/commits/dc70e65553d24337de9178339de8b08100e19eee